### PR TITLE
initialize, open and close metrics client

### PR DESCRIPTION
**Pull request by avsek477**
Depends on: 
- https://github.com/bitfinexcom/bfx-hf-algo/pull/210
- https://github.com/bitfinexcom/bfx-hf-metrics-server/pull/2
- https://github.com/bitfinexcom/bfx-hf-strategy/pull/26
- https://github.com/bitfinexcom/bfx-hf-strategy-exec/pull/24

Metrics server integration.

Returns the following response upon successful authentication.
`['info.auid', auid, {mode: 'main', scope: 'app'}]`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,7 +22,7 @@
     "bfx-api-node-core": "^1.5.8",
     "bfx-api-node-plugin-wd": "^1.0.4",
     "bfx-api-node-rest": "^4.3.1",
-    "bfx-hf-algo": "git+https://github.com/bitfinexcom/bfx-hf-algo.git#v4.13.1",
+    "bfx-hf-algo": "git+https://github.com/avsek477/bfx-hf-algo.git#add-meta-in-algo-orders",
```
Comment (avsek477) [inline package.json:25]: needs to be changed after merge: https://github.com/bitfinexcom/bfx-hf-algo/pull/210

File: package.json [hunk 2]
```diff
@@ -27,7 +27,7 @@
     "bfx-hf-indicators": "git+https://github.com/bitfinexcom/bfx-hf-indicators.git#v2.1.1",
     "bfx-hf-models": "git+https://github.com/bitfinexcom/bfx-hf-models.git#v2.6.0",
     "bfx-hf-models-adapter-lowdb": "git+https://github.com/bitfinexcom/bfx-hf-models-adapter-lowdb.git#v1.0.5",
-    "bfx-hf-strategy": "git+https://github.com/bitfinexcom/bfx-hf-strategy.git#v1.3.1",
+    "bfx-hf-strategy": "git+https://github.com/avsek477/bfx-hf-strategy.git#add-strategy-property-in-meta",
```
Comment (avsek477) [inline package.json:30]: wait for merge: https://github.com/bitfinexcom/bfx-hf-strategy/pull/26

File: package.json [hunk 3]
```diff
@@ -28,7 +28,7 @@
     "bfx-hf-models": "git+https://github.com/bitfinexcom/bfx-hf-models.git#v2.6.0",
     "bfx-hf-models-adapter-lowdb": "git+https://github.com/bitfinexcom/bfx-hf-models-adapter-lowdb.git#v1.0.5",
     "bfx-hf-strategy": "git+https://github.com/avsek477/bfx-hf-strategy.git#add-strategy-property-in-meta",
-    "bfx-hf-strategy-exec": "git+https://github.com/bitfinexcom/bfx-hf-strategy-exec.git#v1.3.1",
+    "bfx-hf-strategy-exec": "git+https://github.com/avsek477/bfx-hf-strategy-exec.git#strategy-property-in-meta-fork",
```
Comment (avsek477) [inline package.json:31]: wait for merge: https://github.com/bitfinexcom/bfx-hf-strategy-exec/pull/24

File: lib/ws_servers/api/session.js [hunk 4]
```diff
@@ -102,6 +110,13 @@ class Session {
     this.services[this.mode].strategyManager = strategyManager
   }
 
+  /**
+   * @param {StrategyManager} strategyManager
```
Comment (tarcisiozf) [inline lib/ws_servers/api/session.js:114]: docs doesn't match

File: lib/ws_servers/api/session.js [hunk 5]
```diff
@@ -115,6 +130,15 @@ class Session {
     this.credentials[mode] = { apiKey, apiSecret }
   }
 
+  sendDataToMetricsServer (data) {
+    return (data) => {
```
Comment (tarcisiozf) [inline lib/ws_servers/api/session.js:134]: this `data` args overrides the one above.

File: lib/ws_servers/api/start_connections.js [hunk 6]
```diff
@@ -115,6 +118,28 @@ class ConnectionManager {
...
+   * @param {boolean} shouldUpdate
+   * @private
+   */
+  async _startMetricsClient (server, session, ws, shouldUpdate = false) {
+    let metricsClient = session.getMetricsClient()
+    if (metricsClient && metricsClient.isOpen && !shouldUpdate) {
+      return { metricsClient: true }
+    }
+
+    metricsClient = await createMetricsClient(server, ws)
```
Comment (tarcisiozf) [inline lib/ws_servers/api/start_connections.js:134]: await is not required

File: lib/ws_servers/api/metrics_client.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+const WebSocket = require('ws')
+const flatPromise = require('flat-promise')
+const { RESTv2 } = require('bfx-api-node-rest')
+
+const maxWaitTime = 2 * 1000
+const pingInterval = 20 * 1000
+const delayReconnect = 1000
+const tokenTTL = 60 * 60
+
+const DEFAULT_METRICS_SERVER_URL = 'ws://localhost:9090/ws/'
```
Comment (tarcisiozf) [inline lib/ws_servers/api/metrics_client.js:12]: I rather throw an exception than use a default localhost URL

**Review by tarcisiozf — CHANGES_REQUESTED** [review]
looks good, left some comments

File: lib/ws_servers/api/session.js (hunk 5 above)
Comment (avsek477) [reply lib/ws_servers/api/session.js:134]: the data arg above is not required.. forgot to remove it later. thanks!!

**Review by tarcisiozf — APPROVED** [review]

File: lib/ws_servers/api/session.js [hunk 8]
```diff
@@ -114,7 +120,14 @@ class Session {
    * @param {MetricsClient} metricsClient
    */
   setMetricsClient (metricsClient) {
-    this.services[this.mode].metricsClient = metricsClient
+    if (this.metricsClient) {
```
Comment (avsek477) [inline lib/ws_servers/api/session.js:123]: also, need to add a condition in `closeServices` method to close the metrics client if it's available and open.
`if(this.metricsClient && this.metricsClient.isOpen) this.metricsClient.close()`

Comment (tarcisiozf) [reply lib/ws_servers/api/session.js:123]: thanks for catching that

**Pull request merged**